test/jcpsantiago/bulgogi_test.clj

```diff
@@ -1,9 +1,90 @@
 (ns jcpsantiago.bulgogi-test
   (:require
     [clojure.test :refer :all]
-    [jcpsantiago.bulgogi :refer :all]))
+    [jcpsantiago.bulgogi :as SUT]
+    [clojure.string :as s]))
 
+(defn boolean->int
+  "Cast a boolean to 1/0 integer indicator"
+  [b]
+  (when boolean? b
+        (if (true? b) 1 0)))
 
-(deftest a-test
-  (testing "will be gone soon :D"
-    (is (= 1 1))))
+
+(defn email-name
+  "Lower-cased name of an email address (the bit before @)"
+  [email]
+  (-> email
+      s/lower-case
+      (s/replace-first #"@.*" "")))
+
+
+(defn n-digits-in-email-name
+  "Number of digits in the email name"
+  [{email :email}]
+  ;; we get dependency management for free because everything is just functions
+  (->> (email-name email)
+       (re-seq #"\d")
+       count))
+
+
+(defn n-chars-in-email-mail
+  "Number of characters in the email name i.e. length of the email name"
+  [{email :email}]
+  (-> (email-name email)
+      count))
+
+
+(defn diff-eur-previous-order
+  "Difference in euros between the current order and the previous one."
+  [{current-amount :current-amount previous-amount :previous-amount}]
+  (- current-amount previous-amount))
+
+
+(defn risky-item?
+  "Boolean depending on whether an item is risky or not"
+  [{brand :brand}]
+  (->> brand
+       s/lower-case
+       (re-seq #"baz corp")
+       some?
+       boolean->int))
+
+
+(defn contains-risky-item
+  "Indicator 1/0 depending on whether a risky item is present in the cart"
+  [{items :items}]
+  (->> items
+       (map #(risky-item? %))
+       (some #(= 1 %))
+       boolean->int))
+
+(defn ^{:bulgogi/co-effect ::added} needs-co-effect
+  [{data ::added}]
+  data)
+
+(defn added [_]
+  {::added "some data"})
```

Comment:
also just realised this solution would make it mandatory that functions always output a map with a namespaced keyword, right? I think it's a good idea, makes all functions have "the same type"

Reply:
Nope. Coeffect functions have to return a map with a namespaced keyword because their result gets merged into the input-data map. As a convention, the keyword should match their name, so that anyone relying on the coeffect already knows where the data is found.

```clojure
(defn ^{:bulgogi/co-effect ::added} needs-co-effect
  [{data ::added}] ;; the ::added thing here comes from the ::added coeffect
    data)

(defn added [_]
  {::added "some data"}) ;; this gets merged into the :input-data map
```

Feature functions can still return whatever they want.

Reply:
I tried making a pair myself, works well. The only thing I'm not a super fan of is the fact that, if there is a different signature between a "feature" and a "coeffect", a data scientist using Bulgogi would need to know that that's how the system works and figure out if they are building a feature or a coeffect (I agree this should be obvious in most cases), additionally it's possible for this to evolve over time and feature become coeffects. 

I'm just speculating here :D right now there are 0 users for Bulgogi, so I think it's ok to assume a hypothetical user would know this is how it works.

Reply:
I think you might be right here. In re-frame it's not as confusing because you add one by calling `reg-cofx` and the final handler with `reg-event-fx`. Bulgogi just registers them magically so it's harder to know what you are building.

Conceptually a coeffect is supposed to enrich the data with some extra data, potentially with extra sideeffects. Feature handlers are supposed to just return a calculated feature of the input. But I see how it could be confusing.

I'll take some hammock time to think about just combining them into the same thing. The only negative that I see immediately is that we would have to solve the dependency graph problem from above, but that's not the end of the world.